ci/lint/06_script.sh

```diff
@@ -10,6 +10,8 @@ GIT_HEAD=$(git rev-parse HEAD)
 if [ -n "$CIRRUS_PR" ]; then
   COMMIT_RANGE="${CIRRUS_BASE_SHA}..$GIT_HEAD"
   test/lint/commit-script-check.sh "$COMMIT_RANGE"
+else
+  COMMIT_RANGE="$(git fetch "$CIRRUS_REPO_CLONE_URL" "${CIRRUS_LAST_GREEN_CHANGE:-$CIRRUS_DEFAULT_BRANCH}" && git merge-base FETCH_HEAD HEAD)..$GIT_HEAD"
```

Comment:
wouldn't it be better to skip the check if this is not a pull request? I presume this already happens when pushing to `master`?

Reply:
Could you clarify please what code change you are suggesting?

Reply:
I don't know the code changes needed, but it seems the check currently only runs on pull requests against the base commit.

On master it is presumably skipped because the base commit is equal to the HEAD commit?

So maybe all places where `COMMIT_RANGE` is used can be guarded by a check on `CIRRUS_PR`? Or instead skip if the range is empty?

Reply:
> it seems the check currently only runs on pull requests against the base commit.

`ci/lint/06_script.sh` runs always except cases defined in the `FILTER_TEMPLATE`.

> On master it is presumably skipped because the base commit is equal to the HEAD commit?

According to Cirrus CI [docs](https://cirrus-ci.org/guide/writing-tasks/#environment-variables), the `CIRRUS_BASE_SHA` variable is defined for PRs only. Therefore, it is natural to have distinct code paths to evaluate `COMMIT_RANGE` for PRs and non-PRs.

> So maybe all places where `COMMIT_RANGE` is used can be guarded by a check on `CIRRUS_PR`?

Given `COMMIT_RANGE` is used in two python scripts, it's okay to evaluate it before `test/lint/lint-all.py`. That is the case for the current `ci/lint/06_script.sh` file.

> Or instead skip if the range is empty?

I cannot figure out such cases.

Reply:
Yes, I understand. My point is that there is no point in running COMMIT_RANGE tests on branches, only on PRs

Reply:
I'd keep `lint-git-commit-check.py` for master and release branches, no?

Reply:
Well, for a pull request it is trivial to see which commits to check. For a branch not so.

Does it check all commits ...
* in the history? 
* since the last push?
* since the last green push?
* against maser's merge base?
* ...

Reply:
Why would the default branch be available on cirrus if the ci runs on another branch?

An alternative would be to run it from the last merge `git log --merges -1`

Reply:
> Why would the default branch be available on cirrus if the ci runs on another branch?

Because "the default branch" is a [property](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/changing-the-default-branch) of a GitHub's repository.

See evidences for availability of the `CIRRUS_DEFAULT_BRANCH` variable for any branch:
- https://cirrus-ci.com/task/6384595250708480?logs=print_env#L29
- https://cirrus-ci.com/task/4992732983721984?logs=print_env#L29

Reply:
Sorry I meant the in the git tree, not the environment variable

Reply:
```
fatal: Not a valid object name master
```

https://github.com/bitcoin/bitcoin/runs/8664441400

Reply:
The `CIRRUS_DEFAULT_BRANCH` is been used in the `git fetch` command, which should succeed regardless of the current branch in the git tree.

CI fails in the 24.x branch because the `COMMIT_RANGE` variable is not set at all, and the failed python scripts use the hardcoded "master" name of a branch. For example: https://github.com/bitcoin/bitcoin/blob/ba48fcf4a40c5b9888459511fb4233a1b89184cc/test/lint/lint-whitespace.py#L92-L101

Reply:
> I believe, such an approach won't leave un-linted commits in the commit history.

yes, maybe.

However, I still don't understand why this is needed.

If you run the linters on a branch, it is too late, because force pushing is not allowed.

For example, it is not possible to rewrite the commit subject line if `test/lint/lint-git-commit-check.py` fails on a commit in a non-pull-request branch. (Same for whitespace from `test/lint/lint-whitespace.py`, or the eval from `test/lint/commit-script-check.sh`). There are no other uses of `COMMIT_RANGE` other than those 3.

Reply:
If we didn't sanity check `verify-commits.py` on branches, I would have already completely disabled the linters for branches.

Reply:
> If you run the linters on a branch, it is too late, because force pushing is not allowed.
> 
> For example, it is not possible to rewrite the commit subject line if `test/lint/lint-git-commit-check.py` fails on a commit in a non-pull-request branch. (Same for whitespace from `test/lint/lint-whitespace.py`, or the eval from `test/lint/commit-script-check.sh`). There are no other uses of `COMMIT_RANGE` other than those 3.

There is still a use case when running CI in personal repos though.

> I would have already completely disabled the linters for branches.

That sounds reasonable. But again, this change allows to use the linter CI task in personal repos.

Reply:
Can you explain that use case? It seems fragile to support that, given that `CIRRUS_LAST_GREEN_CHANGE` might "accidentally" be green for "broken" commits (for example if the CI was skipped, etc...). Or `CIRRUS_DEFAULT_BRANCH` might be too far ahead, so that commits are "skipped".

So to summarize, I fail to see the use case, and even if there was a use case, it would be impossible to support it.